### PR TITLE
[FLINK-11458][checkpointing, rest] Add TERMINATE/SUSPEND Job with Savepoint (FLIP-34)

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This is a first (almost complete) implementation of the FLIP-34 effort. FLIP-34 can be found here https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103090212 and proposes the addition of the `suspend` and `drain` commands to the `CLI` and the `REST` API in general.

With this addition, the user will be able to write to the CLI:
`flink stop -s <savepoint_dir> JOB_ID` for suspending a job with a savepoint, and
`flink stop -s <savepoint_dir> -d JOB_ID` from draining the pipeline.

## Brief change log

Not all changes/additions can be described here. A list of the most important changes follows:
* Adds the `CheckpointType.SYNC_SAVEPOINT` for synchronous savepoints
* Adds the `SynchronousSavepointLatch` which blocks the checkpointing thread of the `Task` until also the `notifyCheckpointComplete` is successfully executed. This will guarantee that in the case of a synchronous checkpoint, the task will wait for the notify callback to be completed, before the task can resume execution.
* Adds the `advanceToEndOfTime` and propagates it from the JM and the `CheckpointCoordinator` to the TM.
* Adds the adequate REST commands and handlers to expose this functionality through the REST API.

## Verifying this change

This change added multiple tests and ITCases. 
In addition, this addition was manually verified locally.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know) (adds the new `suspend` and `terminate` commands.
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented yet**)

## NOTE TO REVIEWERS

These changes have not been tested yet with:
 - The `AsyncIO` operations.
 - The `FileMonitoringSource`.
 - Any other operator/function that does not strictly respect the contract of the exposed APIs, e.g. any sink that does asynchronous work in the `notifyOnCheckpointComplete` and does not wait for its completion.

The REST API documentation has to be re-generated as soon as this PR gets accepted.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointTriggerHeaders.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+		return URL;
+	}
+
+	public static StopWithSavepointTriggerHeaders getInstance() {
+		return INSTANCE;
+	}
+
+	@Override
+	public String getDescription() {
+		return "Triggers a savepoint, and optionally cancels the job afterwards.";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointTriggerHeaders.java]: out-dated? (cancel vs stop, stop isn't optional, missing explanation of `end-of-time` field)

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.Nullable;
+
+/**
+ * {@link RequestBody} for stopping a job with a savepoint.
+ */
+public class StopWithSavepointRequestBody implements RequestBody {
+
+	public static final String FIELD_NAME_TARGET_DIRECTORY = "target-directory";
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java]: please use camelCase for fields.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java [hunk 3]
```diff
@@ -49,7 +49,7 @@ public String getDescription() {
 	}
 
 	/**
-	 * Supported termination modes.
+	 * Supported termination modes. todo this will be removed
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java]: should be retained since the existing cancel functionality remains accessible. In case of stop, as discussed offline, we should return a meaningful error to the user about this operation being no longer supported along with an error code that is not 404 (not found); possibly 410 (permanently removed) or 501 (not implemented).

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+		this.targetDirectory = targetDirectory;
+		this.endOfTime = endOfTime;
+	}
+
+	@Nullable
+	public String getTargetDirectory() {
+		return targetDirectory;
+	}
+
+	public boolean advanceToEndOfTime() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java]: why is this one not called `shouldAdvanceToEndOfTime` like in the stop options?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/savepoints/SavepointHandlers.java [hunk 5]
```diff
@@ -101,6 +104,46 @@ public SavepointHandlers(@Nullable final String defaultSavepointDir) {
...
+
+		@Override
+		protected CompletableFuture<String> triggerOperation(
+				final HandlerRequest<StopWithSavepointRequestBody, StopWithSavepointMessageParameters> request,
+				final RestfulGateway gateway) throws RestHandlerException {
+
+			final JobID jobId = request.getPathParameter(JobIDPathParameter.class);
+			final String requestedTargetDirectory = request.getRequestBody().getTargetDirectory();
+
+			if (requestedTargetDirectory == null && defaultSavepointDir == null) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/savepoints/SavepointHandlers.java]: could deduplicate this block; actually the majority of this class could be moved into a shared `AbstractSavepointTriggerHandler` class

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/savepoints/StopWithSavepointHandlersTest.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+		savepointStatusHandler = savepointHandlers.new SavepointStatusHandler(
+				leaderRetriever,
+				TIMEOUT,
+				Collections.emptyMap());
+	}
+
+	@Test
+	public void testSavepointCompletedSuccessfully() throws Exception {
+		final TestingRestfulGateway testingRestfulGateway = new TestingRestfulGateway.Builder()
+				.setStopWithSavepointFunction((JobID jobId, String targetDirectory) -> CompletableFuture.completedFuture(COMPLETED_SAVEPOINT_EXTERNAL_POINTER))
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/savepoints/StopWithSavepointHandlersTest.java]: isn't this twice the usual indentation?

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/savepoints/StopWithSavepointHandlersTest.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Test for the {@link org.apache.flink.runtime.rest.handler.job.savepoints.SavepointHandlers.StopWithSavepointHandler}.
+ *
+ * <p>Shamelessly copied from {@link SavepointHandlersTest}.
+ */
+public class StopWithSavepointHandlersTest {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/savepoints/StopWithSavepointHandlersTest.java]: extends `TstLogger´

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java (hunk 2 above)
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/stop/StopWithSavepointRequestBody.java]: Ok, but if this is the convention, then the `SavepointTriggerRequestBody` should also be updated.

**Comment by kl0u** [comment]
@zentol and @aljoscha I addressed your comments. Please have a look and let me know what you think.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 8]
```diff
@@ -1308,7 +1313,21 @@ private void executeAsyncCallRunnable(Runnable runnable, String callName) {
 			if (executor == null) {
 				// first time use, initialize
 				checkState(userCodeClassLoader != null, "userCodeClassLoader must not be null");
-				executor = Executors.newSingleThreadExecutor(
+
+				// under normal execution, we expect that one thread will suffice, this is why we
+				// keep the core threads to 1 in the case of a synchronous savepoint, we will block
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]:
```suggestion
				// keep the core threads to 1. In the case of a synchronous savepoint, we will block
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 9]
```diff
@@ -225,6 +228,14 @@ protected StreamTask(
 
 	protected abstract void cancelTask() throws Exception;
 
+	protected void advanceToEndOfEventTime() throws Exception {
```
Comment (aljoscha) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java]: I think it would be good to have comments on these two as well.

**Comment by aljoscha** [comment]
I think this looks very good now! But we still need documentation. 😅

**Comment by kl0u** [comment]
Thanks for the review @aljoscha . I integrated your comments.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+		final int newPoolSize = executor.getCorePoolSize() + activeBlockingCalls;
+
+		// We have to reset the core pool size because (quoted from javadoc):
+		// ``
+		// If there are more than corePoolSize but less than maximumPoolSize threads running,
+		// ** a new thread will be created only if the queue is full **.
+		// ``
+
+		executor.setCorePoolSize(newPoolSize);
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java]: You shouldn't be increasing core pool size:
1. in order to let the `ThreadPoolExecutor` shutdown idling threads more quickly
2. currently `newPoolSize` is initialised from `executor.getCorePoolSize()` so you would exponentially increase the pool size with subsequent `executeBlocking` calls: 1, 2, 4, 7, 11, 16, ...

I think you should just keep the `setCorePoolSize` size constant.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A Thread Pool used by the {@link Task} to launch checkpointing tasks, namely
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java]: I would rephrase the comment to be more general. This could be put on top of the field `private volatile BlockingCallMonitoringThreadPool asyncCallDispatcher` where this class is being used.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java (hunk 10 above)
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java]: You are right about the definition of the `newPoolSize` but we have to update the `corePoolSize` because otherwise the solution will not work, as I write in the comment.

**Comment by kl0u** [comment]
Thanks for the review @pnowojski ! I integrated your comments, so let me know what you think.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java (hunk 11 above)
Comment (kl0u) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/BlockingCallMonitoringThreadPool.java]: Done.

**Comment by kl0u** [comment]
Thanks for +1 @pnowojski . I will wait for Travis and then merge.

**Comment by kl0u** [comment]
Merged.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SynchronousSavepointLatch.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+	void setCheckpointId(final long checkpointId) {
+		if (this.checkpointId == NOT_SET_CHECKPOINT_ID) {
+			this.checkpointId = checkpointId;
+		}
+	}
+
+	boolean blockUntilCheckpointIsAcknowledged() throws Exception {
+		synchronized (synchronizationPoint) {
+			if (isSet() && !isDone()) {
```
Comment (kezhuw) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SynchronousSavepointLatch.java]: This should be `for` due to [spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) ?
